docs/content/docs/ops/batch/blocking_shuffle.md

```diff
@@ -0,0 +1,99 @@
+specific language governing permissions and limitations
+under the License.
+-->
+
+# Blocking Shuffle
+
+## Overview
+
+For bounded jobs, Flink has enhanced the support of batch execution mode in both [DataStream API]({{< ref "docs/dev/datastream/execution_mode" >}}) and [Table / SQL]({{< ref "/docs/dev/table/overview" >}}). In the batch execution mode a new type of shuffle, namely blocking shuffle, is introduced. Unlike pipeline shuffle which exchange data via the network directly, the blocking shuffle first persist data onto some kind of storage, and then the downstream tasks fetch these data via network. Since it does not require the upstream and downstream tasks to run simultaneously, it reduces the overall resources required to execute the bounded job. 
+
+As a whole, Flink provides two different types of blocking shuffles:
+
+- `Hash shuffle`.
+- `Sort shuffle`.
+
+They will be detailed in the following sections.
+
+## Hash Shuffle
+
+The default blocking shuffle implementation is `Hash Shuffle`. In `hash shuffle`, each upstream task would write a separate file for each downstream task on the local disk of the TaskManager. When the downstream tasks run they will request partitions from the upstream TaskManagers, and the upstream TaskManagers will read the data from the files and transmit the data via network.
+
+`Hash Shuffle` provides different mechanisms for writing and reading files:
+
+ - `file`: Writes files with the normal File IO, reads and transmits files with Netty `FileRegion`. `FileRegion` relies on `sendfile` system call to reduce the number of data copies and memory consumption.
+ - `mmap`: Writes and reads files with `mmap` system call.
+ - `Auto`: Writes files with the normal File IO, for file reading, it falls back to normal `file` option on 32 bit machine and use `mmap` on 64 bit machine. This is to avoid file size limitation of java `mmap` implementation on 32 bit machine.
+
+The different mechanism could be chosen via
+
+```yaml
+taskmanager.network.blocking-shuffle.type: file / mmap/ auto
+```
+
+{{< hint info >}}
+**Note:** This option is experimental and might be changed future.
+
+**Note:** If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `file` mechanism could not use `FileRegion` any more and has to use un-pooled buffer to cache data before transmitting. This might [cause direct memory OOM](https://issues.apache.org/jira/browse/FLINK-15981). Also, since the synchronous file reading might blocking Netty thread for some time, the SSL handshake timeout need to be increased to avoid [connection reset errors](https://issues.apache.org/jira/browse/FLINK-21416):
+```yaml
+security.ssl.internal.handshake-timeout: <larger value than the default 10000 ms>
+```
+
+**Note:** The memory usage of `mmap` is not accounted by configured memory limits, but some resource frameworks like yarn would track this memory usage and kill the container once memory exceeding some threshold.
+
+{{< /hint >}}
```

Comment:
```suggestion
{{< hint warning >}}
This option is experimental and might be changed future.
{{< /hint >}}

{{< hint warning >}}
If [SSL]({{< ref "docs/deployment/security/security-ssl" >}}) is enabled, the `file` mechanism can not use `FileRegion` and instead uses an un-pooled buffer to cache data before transmitting. This might [cause direct memory OOM](https://issues.apache.org/jira/browse/FLINK-15981). Additionally, since the synchronous file reading might block Netty threads for some time, the [SSL handshake timeout]({{< ref "docs/deployment/config#security-ssl-internal-handshake-timeout" >}}) needs to be increased to avoid [connection reset errors](https://issues.apache.org/jira/browse/FLINK-21416):
{{< /hint >}}

{{< hint info >}}
The memory usage of `mmap` is not accounted for by configured memory limits, but some resource frameworks like Yarn will track this memory usage and kill the container if memory exceeds some threshold.
{{< /hint >}}
```